core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -821,6 +823,49 @@ private RexNode simplifyIsNull(RexNode a) {
     }
   }
 
+  /**
+   * Validates strong policy for specified {@link RexNode} rex node.
+   *
+   * @param rexNode rex node whose strong policy should be validated.
+   */
+  private void validateStrongPolicy(RexNode rexNode) {
+    if (hasCustomNullabilityRules(rexNode.getKind())) {
+      return;
+    }
+    switch (Strong.policy(rexNode.getKind())) {
+    case NOT_NULL:
+      assert !rexNode.getType().isNullable();
+      break;
+    case ANY:
+      List<RexNode> operands = ((RexCall) rexNode).getOperands();
+      if (rexNode.getType().isNullable()) {
+        assert operands.stream()
+            .map(RexNode::getType)
+            .anyMatch(RelDataType::isNullable);
+      } else {
+        assert operands.stream()
+            .map(RexNode::getType)
+            .noneMatch(RelDataType::isNullable);
+      }
+    }
+  }
+
+  /**
+   * Returns {@code true} if specified {@link SqlKind} sql kind has custom nullability rules.
```

Comment:
IIUC it means that whether expression is null does not only depend on if the input operands contains null, but also depend on its own calculation logic.
@vvysotskyi  shall we doc and explain more here ?

Reply:
Yes, thats correct. I have corrected JavaDoc and provided an example with CAST there.

Reply:
Thanks a lot ~